Restructure offline drivers

micaeljtoliveira opened the pull request:
# CABLE

## Description

Restructure offline drivers such that there is only one executable being build, regardless of MPI support being available or not. Also add the option to run with only one MPI process when MPI support is available (this falls back to the serial driver).

To achieve this, a new mpi_grt_t class is introduced, that hides several calls to the MPI library and takes care of providing the necessary infrastructure when MPI support is not available.

## Type of change

Please delete options that are not relevant.

- [X] Code refactor

## Checklist

- [X] The new content is accessible and located in the appropriate section.
- [X] I have checked that links are valid and point to the intended content.
- [X] I have checked my code/text and corrected any misspellings

<!-- readthedocs-preview cable start -->
----
📚 Documentation preview 📚: https://cable--469.org.readthedocs.build/en/469/

<!-- readthedocs-preview cable end -->

micaeljtoliveira commented on src/offline/cable_driver.F90:
Done.

SeanBryan51 replied:
Can you repeat this process for renaming cable_mpidrv.F90 to cable_driver.F90? Strangely GitHub thinks cable_mpidrv.F90 has been renamed to cable_mpimaster_stub.F90 which I don't think is the intention.

micaeljtoliveira replied:
Okay, done.

micaeljtoliveira commented:
@SeanBryan51 I've now changed the git history and one of the filenames as per our discussion today. Have a look at the changes and let me know if it's better now.

SeanBryan51 commented:
Thanks @micaeljtoliveira. The rendered diff looks a lot better now

SeanBryan51 commented on src/offline/cable_driver_init.F90:
```diff
@@ -84,38 +81,25 @@ MODULE cable_driver_init_mod
 CONTAINS
 
-  SUBROUTINE cable_driver_init()
+  SUBROUTINE cable_driver_init(mpi_grp)
+    TYPE(mpi_grp_t), INTENT(IN) :: mpi_grp
+
     !! Model initialisation routine for the CABLE offline driver.
```
Please describe `mpi_grp` in the ford docstring. See https://cable--469.org.readthedocs.build/en/469/api/proc/cable_driver_init.html for preview

SeanBryan51 commented on src/offline/cable_mpimaster_stub.F90:
```diff
@@ -0,0 +1,32 @@
+
+CONTAINS
+
+  SUBROUTINE mpidrv_master(comm)
```
Please add a docstring for the stub:
```suggestion
  SUBROUTINE mpidrv_master(comm)
    !! <docstring>
```
Preview: https://cable--469.org.readthedocs.build/en/469/api/proc/mpidrv_master~2.html

micaeljtoliveira replied:
Done.

SeanBryan51 commented on src/offline/cable_mpiworker_stub.F90:
```diff
@@ -0,0 +1,32 @@
+!==============================================================================
+! This source code is part of the
+! Australian Community Atmosphere Biosphere Land Exchange (CABLE) model.
+! This work is licensed under the CSIRO Open Source Software License
+! Agreement (variation of the BSD / MIT License).
+!
+! You may not use this file except in compliance with this License.
+! A copy of the License (CSIRO_BSD_MIT_License_v2.0_CABLE.txt) is located
+! in each directory containing CABLE code.
+!
```
See earlier related comment on copyright statement.

micaeljtoliveira replied:
Done.

SeanBryan51 commented on src/offline/cable_mpiworker_stub.F90:
```diff
@@ -0,0 +1,32 @@
+! A copy of the License (CSIRO_BSD_MIT_License_v2.0_CABLE.txt) is located
+! in each directory containing CABLE code.
+!
+! ==============================================================================
+! Purpose: Stub for the worker driver when MPI is not available.
+!
+! ==============================================================================
+!
+MODULE cable_mpiworker
```
Render description as ford module docstring
```suggestion
MODULE cable_mpiworker
  !! Stub for the worker driver when MPI is not available.
```
Preview: https://cable--469.org.readthedocs.build/en/469/api/module/cable_mpiworker.html

SeanBryan51 commented on src/offline/cable_mpiworker_stub.F90:
```diff
@@ -0,0 +1,32 @@
+
+CONTAINS
+
+  SUBROUTINE mpidrv_worker(comm)
```
Please add a docstring for the stub:
```suggestion
  SUBROUTINE mpidrv_worker(comm)
    !! <docstring>
```
Preview: https://cable--469.org.readthedocs.build/en/469/api/proc/mpidrv_worker.html

micaeljtoliveira replied:
Done.

SeanBryan51 commented on src/offline/cable_offline_driver.F90:
```diff
@@ -8,40 +8,36 @@
 ! A copy of the License (CSIRO_BSD_MIT_License_v2.0_CABLE.txt) is located
 ! in each directory containing CABLE code.
 !
```
See earlier related comment on copyright statement.

micaeljtoliveira replied:
Done.

SeanBryan51 commented on src/offline/cable_serial.F90:
```diff
@@ -58,7 +58,7 @@
 !==============================================================================
 
-PROGRAM cable_offline_driver
+MODULE cable_serial
```
Please add module docstring:
```suggestion
MODULE cable_serial
  !! <docstring>
```
Preview: https://cable--469.org.readthedocs.build/en/469/api/module/cable_serial.html

micaeljtoliveira replied:
Done.

SeanBryan51 commented on src/offline/cable_serial.F90:
```diff
@@ -146,6 +146,14 @@ PROGRAM cable_offline_driver
+
+CONTAINS
+
+SUBROUTINE serialdrv()
```
Please add docstring:
```suggestion
SUBROUTINE serialdrv()
  !! <docstring>
```
Preview: https://cable--469.org.readthedocs.build/en/469/api/proc/serialdrv.html

micaeljtoliveira replied:
Done.

SeanBryan51 commented on src/offline/cable_serial.F90:
Please search for and update all occurrences of `cable_driver.F90` in the repo

micaeljtoliveira replied:
Should be done.

SeanBryan51 commented on CMakeLists.txt:
```diff
@@ -150,29 +153,30 @@ add_library(
         cable-mpi
-        src/offline/cable_driver_init.F90
         src/offline/cable_mpicommon.F90
-        src/offline/cable_mpidrv.F90
         src/offline/cable_mpimaster.F90
         src/offline/cable_mpiworker.F90
         src/science/pop/pop_mpi.F90
-        "$<TARGET_OBJECTS:cable_common_objlib>"
+        src/offline/cable_offline_driver.F90
     )
-    target_compile_definitions(cable-mpi PRIVATE __MPI__)
-    target_link_libraries(cable-mpi PRIVATE PkgConfig::NETCDF MPI::MPI_Fortran)
+    target_link_libraries(cable-mpi PRIVATE cable_common MPI::MPI_Fortran)
     install(TARGETS cable-mpi RUNTIME)
+else()
+    add_executable(
+        cable
+        src/offline/cable_mpimaster_stub.F90
+        src/offline/cable_mpiworker_stub.F90
+        src/offline/cable_offline_driver.F90
+    )
+    target_link_libraries(cable PRIVATE cable_common)
+    install(TARGETS cable RUNTIME)
 endif()
```
As discussed, this will break benchcab for the case both `cable` and `cable-mpi` executables are needed as benchcab will only invoke `cmake` once to build both executables. However benchcab will still work with these changes for serial only (`benchcab fluxsite`) or mpi only (`benchcab spatial`) tests (FYI @ccarouge).

It would be good to think about the proper CMake workflow for building different variants of the cable executable (e.g. MPI on/off, Debug, Release, compiler). [From reading online](https://cmake.org/cmake/help/latest/guide/tutorial/Packaging%20Debug%20and%20Release.html#:~:text=By%20default%2C%20CMake%27s%20model%20is%20that%20a%20build%20directory%20only%20contains%20a%20single%20configuration%2C%20be%20it%20Debug%2C%20Release%2C%20MinSizeRel%2C%20or%20RelWithDebInfo.), it seems the typical approach for doing this is to generate separate build trees for each variant unless we work with a multiconfig build system.

micaeljtoliveira replied:
> it seems the typical approach for doing this is to generate separate build trees for each variant

Yes, I think that's the usual strategy. CABLE is quite fast to compile, so this should not be an issue.

One thing you might want to explore to make this process easier is the use of CMake presets. These are really nice.

SeanBryan51 replied:
See https://github.com/CABLE-LSM/CABLE/issues/485

SeanBryan51 reviewed (changes requested):
Nice work @micaeljtoliveira - cool to see Fortran OOP features in action for the MPI data types! I'm happy with all the functional changes you have put in here.

I've been a bit ruthless in my review on documenting things with Ford but I think its good to set a standard for everyone to make sure things are documented properly.

micaeljtoliveira commented on src/offline/cable_driver_init.F90:
```diff
@@ -84,38 +81,25 @@ MODULE cable_driver_init_mod
+  SUBROUTINE cable_driver_init(mpi_grp)
+    TYPE(mpi_grp_t), INTENT(IN) :: mpi_grp
+
     !! Model initialisation routine for the CABLE offline driver.
```
Done.

micaeljtoliveira commented on src/offline/cable_mpi.F90:
```diff
@@ -0,0 +1,160 @@
+
+  INTEGER, PARAMETER :: MPI_COMM_UNDEFINED = -1
+
+  INTEGER :: default_comm ! Deafult communicator to use when creating groups
```
Fixed.

micaeljtoliveira commented on src/offline/cable_mpi.F90:
```diff
@@ -0,0 +1,160 @@
+! A copy of the License (CSIRO_BSD_MIT_License_v2.0_CABLE.txt) is located
+! in each directory containing CABLE code.
+
+MODULE cable_mpi_mod
```
Done.

micaeljtoliveira commented on src/offline/cable_mpi.F90:
```diff
@@ -0,0 +1,160 @@
+    ! the current proccess.
+    INTEGER :: comm = MPI_COMM_UNDEFINED  ! Communicator
+    INTEGER :: rank = -1   ! Rank of the current process
+    INTEGER :: size = -1   ! Size of the communicator
```
Done.

micaeljtoliveira commented on src/offline/cable_mpi.F90:
```diff
@@ -0,0 +1,160 @@
+    !
+    ! Note that when the undefined communicator is used, the group size is 1 and
+    ! the rank to 0, such that the code can work in serial mode.
+    INTEGER, INTENT(IN), OPTIONAL :: comm
```
Done.

micaeljtoliveira commented on src/offline/cable_mpi.F90:
```diff
@@ -0,0 +1,160 @@
+  SUBROUTINE mpi_check_error(ierr)
+    !* Check if an MPI return code signaled an error. If so, print the
+    ! corresponding message and abort the execution.
+    INTEGER, INTENT(IN) :: ierr
```
Done.

micaeljtoliveira commented on src/offline/cable_mpimaster_stub.F90:
```diff
@@ -0,0 +1,32 @@
+! A copy of the License (CSIRO_BSD_MIT_License_v2.0_CABLE.txt) is located
+! in each directory containing CABLE code.
+!
+! ==============================================================================
```
Done.

micaeljtoliveira commented on src/offline/cable_mpimaster_stub.F90:
```diff
@@ -0,0 +1,32 @@
+!
+! ==============================================================================
+!
+MODULE cable_mpimaster
```
Done.

micaeljtoliveira commented:
> I've been a bit ruthless in my review on documenting things with Ford but I think its good to set a standard for everyone to make sure things are documented properly.

No worries. I think I've now addressed all your comments. Please have a look and let me know if I missed something.

ccarouge reviewed (approved):
I just have a question about the unified approach. What happens if someone launches the serial executable via mpirun using several processors by mistake? Wasted resources since it will run on 1 proc but "charge" the resources of several?

SeanBryan51 commented:
> What happens if someone launches the serial executable via mpirun using several processors by mistake?

Testing the following:
```
module add intel-mpi/2019.5.281
mpiexec -n 4 ./cable
```
Looks like it invokes the serial executable in 4 processes in parallel.

SeanBryan51 commented:
> Looks like it invokes the serial executable in 4 processes in parallel.

I've tested this with `openmpi/4.1.4` and the behaviour is the same. This seems like a good argument to have the MPI executable named differently from the serial executable.

micaeljtoliveira commented on src/coupled/ACCESS-CM2/cable_cbm.F90:
```diff
@@ -15,7 +15,7 @@
 !
-! Called from: cable_driver for offline version
+! Called from: cable_serial for offline version
 !              cable_explicit_driver, cable_implicit_driver for ACCESS
```
I've reverted this change.

micaeljtoliveira commented:
> I just have a question about the unified approach. What happens if someone launches the serial executable via mpirun using several processors by mistake? Wasted resources since it will run on 1 proc but "charge" the resources of several?

At the moment this does not really change the current behavior, as there are still two executables. But it might be a source of confusion if later on there's only one executable. That said, from personal experience, it's usually quite obvious when one launches a non-MPI exec with MPI, as the output is duplicated for each MPI rank, as @SeanBryan51 just confirmed.

SeanBryan51 commented:
I've tested these changes with `benchcab fluxsite` and `benchcab spatial` so these changes are good to go :rocket: :

PBS log file:

```
2024-11-15 15:10:18,283 - INFO - benchcab.benchcab.py:369 - Running comparison tasks...
2024-11-15 15:10:18,307 - INFO - benchcab.benchcab.py:370 - tasks: 168 (models: 2, sites: 42, science configurations: 4)
2024-11-15 15:13:09,010 - INFO - benchcab.benchcab.py:381 - 0 failed, 168 passed
```

Benchcab version: `4.1.0`

`config.yaml`:
```
realisations:
  - repo:
      git:
        branch: main
  - repo:
      git:
        branch: introduce_mpi_group

modules: [
  intel-compiler/2021.1.1,
  netcdf/4.7.4,
  openmpi/4.1.0
]
```

Spatial (MPI) runs all execute as expected (note outputs do not reproduce as per issue #463).
